aptos-move/framework/src/natives/cryptography/algebra/arithmetics/scalar_mul.rs

```diff
@@ -0,0 +1,194 @@
+    abort_unless_group_scalar_mul_enabled!(context, group_opt, scalar_field_opt);
+    match (group_opt, scalar_field_opt) {
+        (Some(Structure::BLS12381G1Affine), Some(Structure::BLS12381Fr)) => {
+            ark_scalar_mul_internal!(
+                context,
+                args,
+                ark_bls12_381::G1Projective,
+                ark_bls12_381::Fr,
+                mul_bigint,
+                gas_params.ark_bls12_381_g1_proj_scalar_mul * NumArgs::one()
+            )
+        },
+        (Some(Structure::BLS12381G2Affine), Some(Structure::BLS12381Fr)) => {
+            ark_scalar_mul_internal!(
+                context,
+                args,
+                ark_bls12_381::G2Projective,
+                ark_bls12_381::Fr,
+                mul_bigint,
+                gas_params.ark_bls12_381_g2_proj_scalar_mul * NumArgs::one()
+            )
+        },
+        (Some(Structure::BLS12381Gt), Some(Structure::BLS12381Fr)) => {
+            let scalar_handle = safely_pop_arg!(args, u64) as usize;
+            let element_handle = safely_pop_arg!(args, u64) as usize;
+            safe_borrow_element!(
+                context,
+                element_handle,
+                ark_bls12_381::Fq12,
+                element_ptr,
+                element
+            );
+            safe_borrow_element!(
+                context,
+                scalar_handle,
+                ark_bls12_381::Fr,
+                scalar_ptr,
+                scalar
+            );
+            let scalar_bigint: ark_ff::BigInteger256 = (*scalar).into();
+            context.charge(gas_params.ark_bls12_381_fq12_pow_u256 * NumArgs::one())?;
+            let new_element = element.pow(scalar_bigint);
+            let new_handle = store_element!(context, new_element);
+            Ok(smallvec![Value::u64(new_handle as u64)])
+        },
+        _ => Err(SafeNativeError::Abort {
+            abort_code: MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        }),
+    }
+}
+
+macro_rules! ark_msm_internal {
+    (
+        $gas_params:expr,
+        $context:expr,
+        $args:ident,
+        $structure:expr,
+        $element_typ:ty,
+        $scalar_typ:ty
+    ) => {{
+        let scalar_handles = safely_pop_arg!($args, Vec<u64>);
+        let element_handles = safely_pop_arg!($args, Vec<u64>);
+        let num_elements = element_handles.len();
+        let num_scalars = scalar_handles.len();
+        if num_elements != num_scalars {
+            return Err(SafeNativeError::Abort {
+                abort_code: MOVE_ABORT_CODE_INPUT_VECTOR_SIZES_NOT_MATCHING,
+            });
+        }
+        let mut bases = Vec::with_capacity(num_elements);
+        for handle in element_handles {
+            safe_borrow_element!(
+                $context,
+                handle as usize,
+                $element_typ,
+                element_ptr,
+                element
+            );
+            bases.push(element.into_affine());
```

Comment:
It's strange that you have to convert back to affine here.

Won't the `VariableBaseMSM::msm` call below convert back from affine to projective?

Reply:
Also, should we charge for this conversion?

Reply:
> It's strange that you have to convert back to affine here.
That's how arkworks API currently works.

Fixed the gas charging.

Reply:
Looks like MSMs benefit someohow from affine $(x,y)$ representations (additions are faster). We'll look into it later.